### PR TITLE
Add comprehensive tests for tuners

### PR DESCRIPTION
**Pull request by ultmaster**
This PR addresses the first point of #1543.

This PR:

1. Check the table of search space support among tuners.
2. Clarify some limitations discovered by the tests.
3. Unify all `optimize_mode` to be `maximize` by default.
4. ~Refactor the "dependency graph" of some tuners.~
5. ~Modify `.gitignore`.~
6. Fix rewrite NAS space behavior on user defined search space.
7. Uniform can accept `low == high` now.

This PR will not:

1. Add tests for advisors.
2. Add precondition checks for tuners.
3. Refactor redundant code in tuners.

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -55,6 +56,8 @@ jobs:
     displayName: 'Install nni toolkit via source code'
   - script: |
       cd test
+      ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)" < /dev/null 2> /dev/null
```
Comment (xuehui1991) [inline azure-pipelines.yml]: why we need this?

File: src/sdk/pynni/tests/assets/search_space.json [hunk 2]
```diff
@@ -0,0 +1,74 @@
+{
```
Comment (xuehui1991) [inline src/sdk/pynni/tests/assets/search_space.json]: It's nice job! In the same time, can we add some code for validate the json file of search space?

File: azure-pipelines.yml (hunk 1 above)
Comment (ultmaster) [reply azure-pipelines.yml]: I'm still having a hard time to pass the UT. I'll notice everybody for review once this PR is actually ready...

File: src/sdk/pynni/tests/assets/search_space.json [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  "choice_str": {
+    "_type": "choice",
+    "_value": ["cat","dog", "elephant", "cow", "sheep", "panda"]
```
Comment (xuehui1991) [inline src/sdk/pynni/tests/assets/search_space.json]: I wonder why this folder name is "assets"?

File: src/sdk/pynni/tests/test_tuner.py [hunk 4]
```diff
@@ -17,107 +17,109 @@
...
-from nni.tuner import Tuner
-from nni.msg_dispatcher import MsgDispatcher
-from nni.utils import extract_scalar_reward
-from io import BytesIO
+import copy
 import json
+import logging
 from unittest import TestCase, main
 
+# TODO: can we make this tidier?
```
Comment (xuehui1991) [inline src/sdk/pynni/tests/test_tuner.py]: remove TODO if you want to merge into master branch.

File: src/sdk/pynni/tests/test_tuner.py [hunk 5]
```diff
@@ -17,107 +17,109 @@
...
+from nni.gridsearch_tuner.gridsearch_tuner import GridSearchTuner
+from nni.hyperopt_tuner.hyperopt_tuner import HyperoptTuner
+from nni.smac_tuner.smac_tuner import SMACTuner
```
Comment (xuehui1991) [inline src/sdk/pynni/tests/test_tuner.py]: only test these tuner?

File: src/sdk/pynni/tests/test_tuner.py [hunk 6]
```diff
@@ -17,107 +17,109 @@
...
+        - [X] generate_multiple_parameters
+        - [ ] import_data
+        - [ ] trial_end
+        - [ ] receive_trial_result
+    """
+
+    def search_space_test_one(self, tuner_factory, search_space):
+        tuner = tuner_factory()
+        self.assertIsInstance(tuner, Tuner)
+        tuner.update_search_space(copy.deepcopy(search_space))  # TODO: why does update search space change dict?
```
Comment (xuehui1991) [inline src/sdk/pynni/tests/test_tuner.py]: remove TODO.

File: src/sdk/pynni/tests/test_tuner.py [hunk 7]
```diff
@@ -17,107 +17,109 @@
...
+        parameters = tuner.generate_multiple_parameters(list(range(0, 50)))
+        logger.info(parameters)
+        if not parameters:  # TODO: not strict
+            raise ValueError("No parameters generated")
```
Comment (xuehui1991) [inline src/sdk/pynni/tests/test_tuner.py]: the error information seems a little confuse?

File: src/sdk/pynni/tests/test_tuner.py [hunk 8]
```diff
@@ -17,107 +17,109 @@
...
+                    self.assertLessEqual(v, item["_value"][1])
+                if item["_type"].startswith("q"):
+                    multiple = v / item["_value"][2]
+                    if item["_value"][0] + EPS < v < item["_value"][1] - EPS:
+                        self.assertAlmostEqual(int(round(multiple) + EPS), multiple)
+                if item["_type"] in ("qlognormal", "lognormal"):
+                    self.assertGreater(v, 0)
+
+    def search_space_test_all(self, tuner_factory, supported_types=None):
+        with open("assets/search_space.json", "r") as fp:
```
Comment (xuehui1991) [inline src/sdk/pynni/tests/test_tuner.py]: can we make this path as a parameter and don't hard code here?

File: src/sdk/pynni/tests/test_tuner.py [hunk 9]
```diff
@@ -17,107 +17,109 @@
...
+            single_keyword = single.split("_")
+            single_search_space = {single: search_space_all[single]}
+            if any([t in single_keyword for t in supported_types]) and "fail" not in single_keyword:
+                try:
+                    # supports this key
+                    self.search_space_test_one(tuner_factory, single_search_space)
+                    full_supported_search_space.update(single_search_space)
+                except Exception as e:
+                    if "undefined" not in single_keyword:
+                        raise  # TODO: fix undefined
```
Comment (xuehui1991) [inline src/sdk/pynni/tests/test_tuner.py]: raise what's kind of error or exception?

File: src/sdk/pynni/tests/test_tuner.py [hunk 10]
```diff
@@ -17,107 +17,109 @@
...
+        self.search_space_test_all(lambda: HyperoptTuner("tpe"))
+
+    def test_random_search(self):
+        self.search_space_test_all(lambda: HyperoptTuner("random_search"))
+
+    def test_anneal(self):
+        self.search_space_test_all(lambda: HyperoptTuner("anneal"))
+
+    def test_smac(self):
+        # TODO: smac seems not clean and generates a lot of temporary files
```
Comment (xuehui1991) [inline src/sdk/pynni/tests/test_tuner.py]: Can you fix these TODO at this PR or next PR?

File: azure-pipelines.yml (hunk 1 above)
Comment (xuehui1991) [reply azure-pipelines.yml]: OK

**Comment by ultmaster** [comment]
It's a draft PR and not ready for review yet.

File: src/sdk/pynni/tests/test_tuner.py (hunk 10 above)
Comment (ultmaster) [reply src/sdk/pynni/tests/test_tuner.py]: In this PR of course. I create this PR mainly to see pipeline results.

File: azure-pipelines.yml (hunk 1 above)
Comment (ultmaster) [reply azure-pipelines.yml]: Dependency chain: SMAC -> swig -> brew

File: src/sdk/pynni/tests/assets/search_space.json (hunk 3 above)
Comment (ultmaster) [reply src/sdk/pynni/tests/assets/search_space.json]: Because this is needed by the tests

File: src/sdk/pynni/tests/test_tuner.py (hunk 8 above)
Comment (ultmaster) [reply src/sdk/pynni/tests/test_tuner.py]: Why do we need it as a parameter? It's just a test...

File: src/sdk/pynni/tests/test_tuner.py (hunk 7 above)
Comment (ultmaster) [reply src/sdk/pynni/tests/test_tuner.py]: For unsupported search space definitions for tuners, the behavior is undefined. Many tuners do not throw exceptions. So in order to conform with `assertRaises`, we throw an exception here. Maybe we will add a precondition check in future, and we will remove this.

File: src/sdk/pynni/tests/assets/search_space.json (hunk 2 above)
Comment (ultmaster) [reply src/sdk/pynni/tests/assets/search_space.json]: Precondition check will not be included in this PR

File: .gitignore [hunk 11]
```diff
@@ -67,5 +70,10 @@ typings/
 __pycache__
 build
 *.egg-info
+setup.pye
+
+# Python virtual env
+venv
```
Comment (QuanluZhang) [inline .gitignore]: why adding venv?

File: docs/en_US/Compressor/AutoCompression.md [hunk 12]
```diff
@@ -1,3 +1,118 @@
 # Automatic Model Compression on NNI
 
-TBD.
\ No newline at end of file
+It's convenient to implement auto model compression with NNI compression and NNI tuners
```
Comment (QuanluZhang) [inline docs/en_US/Compressor/AutoCompression.md]: please remove unrelated changes

**Comment by liuzhe-lz** [comment]
Commit history is too messed for review. Please consider a rebase/force push.

**Comment by chicm-ms** [comment]
Since the name of this PR is "Add comprehensive tests for tuners",  suggest to seperate this part out to create a seperate PR.

File: .gitignore (hunk 11 above)
Comment (ultmaster) [reply .gitignore]: Because it's a natural practice to install dependencies under venv during development.

**Comment by ultmaster** [comment]
@liuzhe-lz @chicm-ms @QuanluZhang I noted that many of "unrelated changes" are dealing with pylint and docstring issues (not thoroughly, obviously). Current fixes on pylint and docstring are likely to cause many conflicts. Suggest deferring the review to the point after pylint/docstring PRs have been merged.

File: docs/en_US/Compressor/AutoCompression.md (hunk 12 above)
Comment (ultmaster) [reply docs/en_US/Compressor/AutoCompression.md]: Merged from master. resolved.

File: docs/en_US/Tuner/BuiltinTuner.md [hunk 13]
```diff
@@ -143,16 +143,18 @@ tuner:
...
 **Installation**
 
-SMAC need to be installed by following command before first use.
+SMAC need to be installed by following command before first use. As a reminder, `swig` is required for SMAC: for Ubuntu `swig` can be installed with `apt`.
 
 ```bash
 nnictl package install --name=SMAC
 ```
 
+**Known limitations:** Currently, the tuner has only been tested and working on Ubuntu.
```
Comment (QuanluZhang) [inline docs/en_US/Tuner/BuiltinTuner.md]: remove

File: docs/en_US/Tutorial/SearchSpaceSpec.md [hunk 14]
```diff
@@ -86,20 +90,24 @@ All types of sampling strategies and their parameter are listed here:
...
-| GP Tuner   | &#10003; | &#10003; | &#10003; | &#10003; | &#10003; | &#10003; |  |  |    |     |
+| GP Tuner   | &#10003; | &#10003; | &#10003; | &#10003; | &#10003; | &#10003; |  |  |  |  |
 
 
 Known Limitations:
 
-* GP Tuner and Metis Tuner support only **numerical values** in search space(`choice` type values can be no-numeraical with other tuners, e.g. string values). Both GP Tuner and Metis Tuner use Gaussian Process Regressor(GPR). GPR make predictions based on a kernel function and the 'distance' between different points, it's hard to get the true distance between no-numerical values.
+* GP Tuner and Metis Tuner support only **numerical values** in search space (`choice` type values can be no-numeraical with other tuners, e.g. string values). Both GP Tuner and Metis Tuner use Gaussian Process Regressor(GPR). GPR make predictions based on a kernel function and the 'distance' between different points, it's hard to get the true distance between no-numerical values.
+
+* SMAC Tuner does not support randint/uniform with only one candidate, e.g., `[0, 1]` or `[0.9, 0.9]`, while others do not have such problem, as long as they support randint/uniform.
```
Comment (QuanluZhang) [inline docs/en_US/Tutorial/SearchSpaceSpec.md]: remove

File: docs/en_US/Tutorial/SearchSpaceSpec.md [hunk 15]
```diff
@@ -86,20 +90,24 @@ All types of sampling strategies and their parameter are listed here:
...
+
+* Currently, many tuners won't throw exceptions for invalid search space. For unsupported search space types, it's also possible that the tuner will ignore the field like nothing happened.
```
Comment (QuanluZhang) [inline docs/en_US/Tutorial/SearchSpaceSpec.md]: remove, use an issue to track

File: src/sdk/pynni/nni/nas_utils.py [hunk 16]
```diff
@@ -265,6 +265,8 @@ def convert_nas_search_space(search_space):
         param search_space: raw search space
         return: the new search space, mutable_layers will be converted into choice
     """
+    if not isinstance(search_space, dict):
+        return search_space
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/nas_utils.py]: in what scenario that search_space is not a dict?

Comment (ultmaster) [reply src/sdk/pynni/nni/nas_utils.py]: Users might write a customized search space, which is a list. Better careful than sorry. This won't be harmful.

File: src/sdk/pynni/tests/test_tuner.py [hunk 17]
```diff
@@ -17,107 +17,172 @@
...
+            search_space_all = json.load(fp)
+        if supported_types is None:
+            supported_types = ["choice", "randint", "uniform", "quniform", "loguniform", "qloguniform",
+                               "normal", "qnormal", "lognormal", "qlognormal"]
+        full_supported_search_space = dict()
+        for single in search_space_all:
+            single_keyword = single.split("_")
+            space = search_space_all[single]
+            expected_fail = not any([t in single_keyword for t in supported_types]) or "fail" in single_keyword
+            if ignore_types is not None and any([t in ignore_types for t in single_keyword]):
```
Comment (QuanluZhang) [inline src/sdk/pynni/tests/test_tuner.py]: what is the relation between `supported_types` and `ignore_types`?

Comment (ultmaster) [reply src/sdk/pynni/tests/test_tuner.py]: Supported types are listed in the table. They are meant to be supported and should be correct. Other than those, all the rest are "unsupported", which are expected to produce ridiculous results or throw some exceptions. However, there are certain types I can't check. For example, generate "normal" using GP Tuner returns successfully and results are fine if we check the range (-inf to +inf), but they make no sense: it's not a normal distribution.

File: test/unittest.sh [hunk 18]
```diff
@@ -18,11 +18,10 @@ export NNI_PLATFORM=unittest
 echo ""
 echo "===========================Testing: nni_sdk==========================="
 cd ${CWD}/../src/sdk/pynni/
+python3 -m pip install --user -r nni/smac_tuner/requirements.txt
 python3 -m unittest discover -v tests
```
Comment (chicm-ms) [inline test/unittest.sh]: This unittest.sh is used by window IT pipeline,  can smac requirements be installed on Windows?

Comment (ultmaster) [reply test/unittest.sh]: SMAC cannot be used on Windows, as specified in documentation. Suggestions?

File: src/sdk/pynni/tests/test_tuner.py (hunk 17 above)
Comment (QuanluZhang) [reply src/sdk/pynni/tests/test_tuner.py]: please add comments for `ignore_types`

File: test/unittest.sh (hunk 18 above)
Comment (chicm-ms) [reply test/unittest.sh]: You can install the packages in pipeline instead of this unittest.sh.   exclude the smac unit test cases on windows.

Comment (chicm-ms) [reply test/unittest.sh]: It seems SMAC is already installed in pipelines by
`
      PATH=$HOME/.local/bin:$PATH nnictl package install --name=SMAC
`
